core/services/ocr2/plugins/ocr2keeper/evm/mercury_lookup.go

```diff
@@ -0,0 +1,343 @@
+package evm
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/patrickmn/go-cache"
+	"github.com/smartcontractkit/ocr2keepers/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
+)
+
+type MercuryLookup struct {
+	feedLabel  string
+	feeds      []string
+	queryLabel string
+	query      *big.Int
+	extraData  []byte
+}
+
+// MercuryMultiResponse TODO guessing on this
+type MercuryMultiResponse struct {
+	ChainlinkBlobs []string `json:"chainlinkBlob"`
+}
+
+type MercuryResponse struct {
+	ChainlinkBlob string `json:"chainlinkBlob"`
+}
+
+type MercuryBytes struct {
+	Index int
+	Error error
+	Bytes []byte
+}
+
+const (
+	RetryDelay   = 500 * time.Millisecond
+	TotalAttempt = 3
+)
+
+// mercuryLookup looks through check upkeep results looking for any that need off chain lookup
+func (r *EvmRegistry) mercuryLookup(ctx context.Context, upkeepResults []types.UpkeepResult) ([]types.UpkeepResult, error) {
+	// return error only if there are errors which stops the process
+	// don't surface Mercury API errors to plugin bc MercuryLookup process should be self-contained
+	for i := range upkeepResults {
```

Comment:
do we plan to do this in parallel across the results in a future ticket?

Reply:
we can certainly do that in the future

Reply:
(y) can you create a ticket to track this